packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,100 @@
+import { createReadStream } from 'node:fs';
+
+import { uploadFileGuard, maxUploadFileSize } from '@logto/schemas';
+import { generateStandardId } from '@logto/shared';
+import etl from 'etl';
+import mime from 'mime-types';
+import unzipper, { type Entry } from 'unzipper';
+import { object, z } from 'zod';
+
+import { EnvSet } from '#src/env-set/index.js';
+import RequestError from '#src/errors/RequestError/index.js';
+import koaGuard from '#src/middleware/koa-guard.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getConsoleLogFromContext } from '#src/utils/console.js';
+import { buildUploadFile } from '#src/utils/storage/index.js';
+import { getTenantId } from '#src/utils/tenant.js';
+
+import { type ManagementApiRouter, type RouterInitArgs } from '../../types.js';
+
+import { checkEntryForDirectoryOrSystemFile, removeZipRootDirectory } from './utils.js';
+
+export default function customUiAssetsRoutes<T extends ManagementApiRouter>(
+  ...[router, { queries }]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+
+      const file = bodyFiles[0];
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      // Check if the zip package contains an `index.html` file in root directory
+      const zipDirectory = await unzipper.Open.file(file.filepath);
+      assertThat(
+        zipDirectory.files.some((entry) => removeZipRootDirectory(entry.path) === 'index.html'),
+        'sign_in_experiences.index_page_not_found'
+      );
+
+      const { storageProviderConfig } = SystemContext.shared;
+      assertThat(storageProviderConfig, 'storage.not_configured');
+
+      const [tenantId] = await getTenantId(ctx.URL);
+      assertThat(tenantId, 'guard.can_not_get_tenant_id');
+
+      const customUiAssetId = generateStandardId(8);
+      const upload = buildUploadFile(storageProviderConfig);
+
+      try {
+        // Unzip the package in memory and upload the file stream to the storage provider
+        await createReadStream(file.filepath)
+          // eslint-disable-next-line new-cap
+          .pipe(unzipper.Parse({ forceStream: true }))
+          .pipe(
+            etl.map(async (entry: Entry) => {
```

Comment:
what's the rationale of adding `etl`?

1. i see only `etl.map` is used while it's [a simple wrapper of `streamz`](https://github.com/ZJONSSON/node-etl/blob/60e7da365898119ec4e51903f7adff08ca715d5e/lib/index.js#L23)
2. does the functionality of `streamz` necessary for our case?

Reply:
It's from the [sample code](https://github.com/ZJONSSON/node-unzipper?tab=readme-ov-file#parse-zip-by-piping-entries-downstream) of "unzipper" package.
The `etl` is used to map the `entry` object in the stream pipe.

I would like to replace it with other techniques if possible.

Reply:
OK, I also see the `etl.map` is a wrapper of streamz's `map` function. So should I just switch to `streamz` instead? 

I need to map the `entry` object in the unzipping stream, filter out unnecessary folders and system files, and then finally upload UI asset files to our storage provider.

Reply:
i'm not sure if the "unzipper" package is required though

Reply:
Removed and use azure functions to unzip instead